File: web/src/services/accountService.ts
import type { SupabaseClient } from "@supabase/supabase-js";
import { clearOutbox } from "./outboxService";

export type AccountDeletionState = "active" | "pending_deletion";

export interface AccountStatus {
  readonly deletionState: AccountDeletionState;
  readonly deletionDeadline: string | null;
  readonly recoveryAvailable: boolean;
}

export interface DeletionConfirmation {
  readonly confirmed: boolean;
  readonly deletionState: AccountDeletionState;
  readonly deletionDeadline: string | null;
  readonly sessionsRevoked: boolean;
}

export interface OperatorExportSnapshot {
  readonly exportedAt: string;
  readonly tasks: readonly unknown[];
  readonly labels: readonly unknown[];
  readonly taskLabels: readonly unknown[];
  readonly comments: readonly unknown[];
  readonly settings: unknown;
}

export class AccountLifecycleError extends Error {
  readonly status: number;
  readonly code?: string;

  constructor(
    message: string,
    status: number,
    code?: string,
    isNetworkError?: boolean,
  ) {
    super(message);
    this.name = "AccountLifecycleError";
    this.status = status;
    this.code = code;
    if (isNetworkError) {
      (this as { isNetworkError?: boolean }).isNetworkError = true;
    }
  }
}

type LifecycleAction = "status" | "request-deletion" | "recover-account";

async function callLifecycleEndpoint(
  client: SupabaseClient,
  action: LifecycleAction,
): Promise<Record<string, unknown>> {
  const { data, error: sessionError } = await client.auth.getSession();
  const session = sessionError ? null : data.session;
  if (!session) {
    throw new AccountLifecycleError(
      "Please sign in to manage your account.",
      401,
      "unauthorized",
    );
  }

  const {
    data: payload,
    error,
    response,
  } = await client.functions.invoke<Record<string, unknown>>(
    "account-lifecycle",
    {
      headers: { Authorization: `Bearer ${session.access_token}` },
      body: { action },
    },
  );

  if (error) {
    if (response instanceof Response) {
      let failureBody: Record<string, unknown> = {};
      try {
        failureBody = (await response.json()) as Record<string, unknown>;
      } catch {
        // Non-JSON error body falls through to a generic message below
      }
      throw new AccountLifecycleError(
        typeof failureBody.error === "string" && failureBody.error
          ? failureBody.error
          : "Account service request failed.",
        response.status,
        typeof failureBody.code === "string" ? failureBody.code : undefined,
      );
    }
    throw new AccountLifecycleError(
      "Network error: unable to reach Cras account services.",
      0,
      "network_error",
      true,
    );
  }

  return payload ?? {};
}

function readString(value: unknown): string | null {
  return typeof value === "string" ? value : null;
}

export async function fetchAccountStatus(
  client: SupabaseClient,
): Promise<AccountStatus> {
  const payload = await callLifecycleEndpoint(client, "status");
  const deletionState =
    payload.deletionState === "pending_deletion"
      ? "pending_deletion"
      : "active";
  return {
    deletionState,
    deletionDeadline: readString(payload.deletionDeadline),
    recoveryAvailable: payload.recoveryAvailable === true,
  };
}

export async function requestAccountDeletion(
  client: SupabaseClient,
): Promise<DeletionConfirmation> {
  const payload = await callLifecycleEndpoint(client, "request-deletion");
  return {
    confirmed: payload.confirmed === true,
    deletionState:
      payload.deletionState === "pending_deletion"
        ? "pending_deletion"
        : "active",
    deletionDeadline: readString(payload.deletionDeadline),
    sessionsRevoked: payload.sessionsRevoked === true,
  };
}

export async function recoverAccount(client: SupabaseClient): Promise<void> {
  await callLifecycleEndpoint(client, "recover-account");
}

export async function generateAccountExport(
  client: SupabaseClient,
): Promise<OperatorExportSnapshot> {
  const { data, error } = await client
    .schema("api")
    .rpc("export_operator_data");

  if (error || typeof data !== "string") {
    throw new Error(
      `Failed to generate account export: ${error?.message ?? "empty response"}`,
    );
  }

  return JSON.parse(data) as OperatorExportSnapshot;
}

export function downloadJsonFile(fileName: string, contents: string): void {
  const blob = new Blob([contents], { type: "application/json" });
  const url = URL.createObjectURL(blob);
  const anchor = document.createElement("a");
  anchor.href = url;
  anchor.download = fileName;
  document.body.appendChild(anchor);
  anchor.click();
  anchor.remove();
  URL.revokeObjectURL(url);
}

export async function downloadAccountExport(
  client: SupabaseClient,
): Promise<void> {
  const snapshot = await generateAccountExport(client);
  const day = snapshot.exportedAt.slice(0, 10) || new Date().toISOString();
  downloadJsonFile(`cras-export-${day}.json`, JSON.stringify(snapshot));
}

const REAUTH_INTENT_KEY = "cras_reauth_intent";

export const REAUTH_INTENT_STORAGE_KEY = REAUTH_INTENT_KEY;

export function stageReauthIntent(userId: string): void {
  try {
    sessionStorage.setItem(REAUTH_INTENT_KEY, userId);
  } catch {
    // Storage unavailable; the flow continues without resume-on-return
  }
}

export function consumeReauthIntent(expectedUserId: string): string | null {
  try {
    const staged = sessionStorage.getItem(REAUTH_INTENT_KEY);
    sessionStorage.removeItem(REAUTH_INTENT_KEY);
    if (staged && staged === expectedUserId) {
      return staged;
    }
  } catch {
    // Storage unavailable
  }
  return null;
}

const CACHED_TIMED_PLAN_TYPE_KEY = "cras_effective_default_timed_plan_type";
const UNSUBMITTED_TASK_INPUT_KEY = "cras_unsubmitted_task_input";

export function clearLocalOperatorData(operatorId: string): void {
  clearOutbox(operatorId);
  try {
    localStorage.removeItem(CACHED_TIMED_PLAN_TYPE_KEY);
    sessionStorage.removeItem(UNSUBMITTED_TASK_INPUT_KEY);
    sessionStorage.removeItem(REAUTH_INTENT_KEY);
  } catch {
    // Ignore storage access errors during wipe
  }
}

File: android/app/src/main/java/com/cras/app/data/AccountService.kt
package com.cras.app.data

import com.cras.app.auth.OperatorSession
import com.cras.app.config.PublicSupabaseConfig
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException

enum class AccountDeletionState(val value: String) {
    ACTIVE("active"),
    PENDING_DELETION("pending_deletion");

    companion object {
        fun fromValue(value: String?): AccountDeletionState = when (value) {
            "pending_deletion" -> PENDING_DELETION
            else -> ACTIVE
        }
    }
}

@Serializable
data class AccountStatus(
    val deletionState: AccountDeletionState,
    val deletionDeadline: String?,
    val recoveryAvailable: Boolean
)

@Serializable
data class DeletionConfirmation(
    val confirmed: Boolean,
    val deletionState: AccountDeletionState,
    val deletionDeadline: String?,
    val sessionsRevoked: Boolean
)

class AccountLifecycleException(
    message: String,
    val statusCode: Int = 0,
    val code: String? = null,
    val isNetworkError: Boolean = false,
    cause: Throwable? = null
) : Exception(message, cause)

interface AccountService {
    suspend fun fetchAccountStatus(session: OperatorSession): AccountStatus
    suspend fun requestAccountDeletion(session: OperatorSession): DeletionConfirmation
    suspend fun recoverAccount(session: OperatorSession)
    suspend fun exportOperatorData(session: OperatorSession): String
}

private fun JsonElement?.readString(): String? {
    if (this == null || this is JsonNull) return null
    if (this is JsonPrimitive) {
        val str = this.contentOrNull
        return if (str == "null" || str.isNullOrEmpty()) null else str
    }
    return null
}

private fun JsonElement?.readBoolean(defaultValue: Boolean = false): Boolean {
    if (this == null || this is JsonNull) return defaultValue
    if (this is JsonPrimitive) {
        return this.booleanOrNull ?: this.contentOrNull?.toBooleanStrictOrNull() ?: defaultValue
    }
    return defaultValue
}

class SupabaseAccountService(
    private val config: PublicSupabaseConfig,
    private val httpClient: OkHttpClient = OkHttpClient(),
    private val json: Json = Json { ignoreUnknownKeys = true }
) : AccountService {

    private fun callLifecycleEndpoint(session: OperatorSession, action: String): JsonObject {
        val endpoint = "${config.url}/functions/v1/account-lifecycle"
        val requestBodyJson = buildJsonObject {
            put("action", action)
        }.toString()

        val request = Request.Builder()
            .url(endpoint)
            .addHeader("apikey", config.publishableKey)
            .addHeader("Authorization", "Bearer ${session.accessToken}")
            .addHeader("Content-Type", "application/json")
            .post(requestBodyJson.toRequestBody("application/json".toMediaType()))
            .build()

        val response = try {
            httpClient.newCall(request).execute()
        } catch (e: IOException) {
            throw AccountLifecycleException(
                message = "Network error: unable to reach Cras account services.",
                statusCode = 0,
                code = "network_error",
                isNetworkError = true,
                cause = e
            )
        }

        return response.use { resp ->
            val responseBody = resp.body?.string() ?: ""
            if (!resp.isSuccessful) {
                var errorMessage = "Account service request failed."
                var errorCode: String? = null
                try {
                    val parsed = json.parseToJsonElement(responseBody).jsonObject
                    parsed["error"]?.readString()?.let { if (it.isNotBlank()) errorMessage = it }
                    parsed["code"]?.readString()?.let { errorCode = it }
                } catch (_: Exception) {
                    // Fallback to default message
                }
                throw AccountLifecycleException(
                    message = errorMessage,
                    statusCode = resp.code,
                    code = errorCode
                )
            }

            if (responseBody.trim().isEmpty() || responseBody.trim() == "null") {
                JsonObject(emptyMap())
            } else {
                try {
                    json.parseToJsonElement(responseBody).jsonObject
                } catch (_: Exception) {
                    JsonObject(emptyMap())
                }
            }
        }
    }

    override suspend fun fetchAccountStatus(session: OperatorSession): AccountStatus {
        val responseObj = callLifecycleEndpoint(session, "status")
        val stateStr = responseObj["deletionState"]?.readString()
            ?: responseObj["deletion_state"]?.readString()
        val deadlineStr = responseObj["deletionDeadline"]?.readString()
            ?: responseObj["deletion_deadline"]?.readString()
        val recoveryAvail = responseObj["recoveryAvailable"]?.readBoolean(false)
            ?: responseObj["recovery_available"]?.readBoolean(false)
            ?: false

        return AccountStatus(
            deletionState = AccountDeletionState.fromValue(stateStr),
            deletionDeadline = deadlineStr,
            recoveryAvailable = recoveryAvail
        )
    }

    override suspend fun requestAccountDeletion(session: OperatorSession): DeletionConfirmation {
        val responseObj = callLifecycleEndpoint(session, "request-deletion")
        val confirmed = responseObj["confirmed"]?.readBoolean(false) ?: false
        val stateStr = responseObj["deletionState"]?.readString()
            ?: responseObj["deletion_state"]?.readString()
        val deadlineStr = responseObj["deletionDeadline"]?.readString()
            ?: responseObj["deletion_deadline"]?.readString()
        val sessionsRevoked = responseObj["sessionsRevoked"]?.readBoolean(false)
            ?: responseObj["sessions_revoked"]?.readBoolean(false)
            ?: false

        return DeletionConfirmation(
            confirmed = confirmed,
            deletionState = AccountDeletionState.fromValue(stateStr),
            deletionDeadline = deadlineStr,
            sessionsRevoked = sessionsRevoked
        )
    }

    override suspend fun recoverAccount(session: OperatorSession) {
        callLifecycleEndpoint(session, "recover-account")
    }

    override suspend fun exportOperatorData(session: OperatorSession): String {
        val endpoint = "${config.url}/rest/v1/rpc/export_operator_data"
        val request = Request.Builder()
            .url(endpoint)
            .addHeader("apikey", config.publishableKey)
            .addHeader("Authorization", "Bearer ${session.accessToken}")
            .addHeader("Content-Type", "application/json")
            .post("{}".toRequestBody("application/json".toMediaType()))
            .build()

        val response = try {
            httpClient.newCall(request).execute()
        } catch (e: IOException) {
            throw AccountLifecycleException(
                message = "Network error: unable to generate account export.",
                statusCode = 0,
                code = "network_error",
                isNetworkError = true,
                cause = e
            )
        }

        return response.use { resp ->
            val responseBody = resp.body?.string() ?: ""
            if (!resp.isSuccessful) {
                throw AccountLifecycleException(
                    message = "Failed to generate account export: ${resp.code} $responseBody",
                    statusCode = resp.code
                )
            }
            if (responseBody.startsWith("\"") && responseBody.endsWith("\"") && responseBody.length >= 2) {
                try {
                    json.decodeFromString<String>(responseBody)
                } catch (_: Exception) {
                    responseBody
                }
            } else {
                responseBody
            }
        }
    }
}
